File: Others/medium/MinHeightBst.py
def minHeightBst(array):
    return construct_bst_aux(array, 0, len(array) - 1)


def construct_bst_aux(array, left, right):
    if left == right:
        return BST(array[left])

    if left + 1 == right:
        root = BST(array[left])
        root.right = BST(array[right])
        return root

    mid = (left + right) // 2

    root = BST(array[mid])

    root.left = construct_bst_aux(array, left, mid - 1)
    root.right = construct_bst_aux(array, mid + 1, right)

    return root


class BST:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    def insert(self, value):
        if value < self.value:
            if self.left is None:
                self.left = BST(value)
            else:
                self.left.insert(value)
        else:
            if self.right is None:
                self.right = BST(value)
            else:
                self.right.insert(value)

File: Golang/medium/min-height-bst.go
package main

func Solve(array []int, left int, right int) *BST {
	// [1, 2, 5, 7, 10, 13, 14, 15, 22]
	// l = 0, r = 8, m = 3, root = 10, lft = [1, 2, 5, 7], [13, 14, 15, 22]
	if right < left {
		return nil
	}

	if left == right {
		return &BST{Value: array[left]}
	}
	if left < right && left == right-1 {
		return &BST{Value: array[left], Right: &BST{Value: array[right]}}
	}

	mid := left + (right-left)/2
	return &BST{Value: array[mid], Left: Solve(array, left, mid-1), Right: Solve(array, mid+1, right)}
}

func MinHeightBST(array []int) *BST {
	return Solve(array, 0, len(array)-1)
}

type BST struct {
	Value int

	Left  *BST
	Right *BST
}
